cmd/composite_gendoc.go

```diff
@@ -0,0 +1,272 @@
+	key string
+	description string
+	required bool
+	valueType string `schema:"type"`
+	defaultValues interface{} `schema:"default"`
+}
+
+func generateConfigContent(configFilePath string) (string, error) {
+
+	filename, err := filepath.Abs(configFilePath)
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(-1)
+	}
+
+	yamlFile, err := ioutil.ReadFile(filename)
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		fmt.Fprintln(os.Stderr, "Could not read "+filename)
+	}
+
+	var config YamlConfig
+
+	err = yaml.Unmarshal(yamlFile, &config)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Could not parse YAML")
+		os.Exit(-1)
+	}
+
+
+	missingRequiredContent, err := generateVariablesContent(
+		config,
+		func(required bool, defaultValue interface{}) bool{
+			return required && defaultValue == nil
+		},
+		fmt.Sprintf("\n## %s\n\n", content.DEFAULT_FILES_GENDOC_README_REQUIRED_NO_DEFAULT_HEADER),
+		false)
+
+	requiredContent, err := generateVariablesContent(
+		config,
+		func(required bool, defaultValue interface{}) bool{
+			return required && defaultValue != nil
+		},
+		fmt.Sprintf("\n## %s\n\n", content.DEFAULT_FILES_GENDOC_README_REQUIRED_DEFAULT_HEADER),
+		true)
+
+	notRequiredDefaultContent, err := generateVariablesContent(
+		config,
+		func(required bool, defaultValue interface{}) bool{
+			return !required && defaultValue != nil
+		},
+		fmt.Sprintf("\n## %s\n\n", content.DEFAULT_FILES_GENDOC_README_NO_REQUIRED_DEFAULT_HEADER),
+		true)
+
+
+	theRestContent, err := generateVariablesContent(
+		config,
+		func(required bool, defaultValue interface{}) bool{
+			return !required && defaultValue == nil
+		},
+		fmt.Sprintf("\n## %s\n\n", content.DEFAULT_FILES_GENDOC_README_NO_REQUIRED_NO_DEFAULT_HEADER),
+		true)
+
+	return missingRequiredContent + requiredContent + notRequiredDefaultContent+theRestContent, err
+
+}
+
+
+func generateVariablesContent(config YamlConfig, check func(bool, interface{}) bool, header string, printVar bool ) (string, error) {
+
+	var contentBytes bytes.Buffer
+	counter := 0
+	contentBytes.WriteString(header)
+
+	// loop over mapslice and create option object. TODO: Better way would be to use reflection instead.
+	for _, variable := range config.Variables {
+		option := varOption{}
+		option.key = variable.Key.(string)
+		for _, o := range variable.Value.(yaml.MapSlice) {
+			if o.Value != nil {
+				switch o.Key {
+				case "description":
+					option.description = o.Value.(string)
+				case "required":
+					option.required = o.Value.(bool)
+				case "type":
+					option.valueType = o.Value.(string)
+				case "default":
+					option.defaultValues = o.Value
+				}
+			}
+		}
+
+		// check if
+		if check(option.required, option.defaultValues) {
+			counter ++
+			contentBytes.WriteString("### `"+ option.key + "`:\n")
+			contentBytes.WriteString("  * description: "+ option.description)
+
+			if printVar && option.defaultValues != nil{
+
+				switch option.valueType  {
+				case "boolean":
```

Comment:
we need the "case" type as well - 
check this out:
https://github.com/CloudCoreo/etcd-cluster/blob/master/config.yaml#L43